src/include/sof/audio/buffer.h

```diff
@@ -271,6 +271,48 @@ static inline uint32_t buffer_avail_frames(struct comp_buffer *source,
 	return MIN(src_frames, sink_frames);
 }
 
+/**
+ * Returns the contiguous memory regions which are free and available
+ * in this buffer.
+ * @param buf Component buffer
+ * @param[out] r_ptr Beginning of readable memory region
+ * @param[out] r_size Size (in bytes) of readable memory region
+ * @param[out] w_ptr Beginning of writeable memory region
+ * @param[out] w_size Size (in bytes) of writeable memory region
+ * @return 0 on success, -EINVAL if buf is NULL or all 4 output parameters are
+ *         NULL
+ */
+static inline int buffer_contig_avail_free(struct comp_buffer *buf,
+	void **r_ptr, size_t *r_size, void **w_ptr, size_t *w_size)
+{
+	char *read;
+	char *write;
+	size_t read_size;
+	size_t write_size;
+
+	if (!buf)
+		return -EINVAL;
+	if (!r_ptr && !r_size && !w_ptr && !w_size)
+		return -EINVAL;
+	read = buf->r_ptr;
+	read_size = buf->avail;
+	if (read + read_size > (char *)buf->end_addr)
+		read_size = (char *)buf->end_addr - read;
```

Comment:
pedantically speaking `char` isn't guaranteed to be 1 byte in size, so, where you do byte operations (like here), it's better to use `int8_t`. Also `read` and `write` can be defined of that type.

Reply:
char is however guaranteed by the C standard to be the minimum addressable unit. sizeof(char) is guaranteed to be 1. If char is 16 bits, int8_t doesn't exist.

I will take a peek in how this was solved in other places in the code and will align if (u)int8_t is used more commonly.

Reply:
Sorry, the rest of the file uses `char *` (I think I have myself to blame here from a not-very-distant update I did with the arithmetic), won't change to `int8_t` for one function alone. Maybe next refactor.